### PR TITLE
Multiple Dispatch

### PR DESCRIPTION
**Pull request by softmattertheory**
This PR adds initial support for typing and multiple dispatch to Morpho. You can define multiple implementations of the same function: 

````
fn f(List x) { .. }
fn f(String x) { .. }
````

and morpho will dispatch to the correct implementation based on the type at runtime. Multiple dispatch resolutions are compiled to a special implementation (a directed acyclic graph) for efficient execution, imposing typically only a 20% overhead on a regular function call per argument. Future compiler optimizations will reduce the need for this. 

Multiple dispatch can similarly be used for method definitions, and are resolved at runtime. PR includes extensive tests and online help.  

Typed variables and support for type tracking in the compiler are implemented, but type checking is not fully implemented yet. (This PR is a necessary step towards full type checking).

### CODE REVIEW

**Comment by joshichaitanya3** [comment]
This looks great! This would both simplify and improve so many core things! I love the improvement on the List_sort API. And we got the visitor pattern in graphics already working with multiple dispatch. Great step for both users and developers!

I have added some comments with small suggested changes / additions. 

One thing I would recommend is adding something like a `types.md` help file where it would list all the in-built types and their details, future supported in-built types (Callable?), and syntax, etc. Since this is a big change and more related features are a WIP, it would be great to have this documented as it would give users an idea of what currently works and what would work in the future. This is not urgent, but would be really helpful if it exists when this feature is released. 

Really excited for this!

**Review by joshichaitanya3 — CHANGES_REQUESTED** [review]

File: help/functions.md [hunk 1]
```diff
@@ -69,6 +89,50 @@ The `return` keyword is used to exit from a function, optionally passing a given
...
+    fn f(List x, y) {} // The first parameter must be a list
+    fn f(List x, List y) {} // Both parameters must be lists
+
+While you can define multiple implementations of a function that accept different parameter types, you can only define one implementation with a unique signature.
+
+Note that optional and variadic parameters are not typed.
+
+# Multiple dispatch
+[tagmultiple]: # (multiple)
+[tagdispatch]: # (dispatch)
```
[the remark this answers is not in the document]
Comment (softmattertheory) [reply help/functions.md:108]: Added

File: test/types/multiple_dispatch/duplicate_no_type.morpho [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// Duplicate implementations with one argument no Type annotations.
+// This one currently segfaults due to an infinite loop.
```
[the remark this answers is not in the document]
Comment (softmattertheory) [reply test/types/multiple_dispatch/duplicate_no_type.morpho:2]: Done!

File: test/types/multiple_dispatch/duplicate_one_arg.morpho [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// Duplicate implementations with a single Typed argument. 
+// This also segfaults similar to `duplicate_no_type.morpho`
+
+fn f(Int x) {
+    return 0
+}
+
+fn f(Int x) {
+    return 1 
+}
```
[the remark this answers is not in the document]
Comment (softmattertheory) [reply test/types/multiple_dispatch/duplicate_one_arg.morpho:10]: Great idea! Added!

**Pull request merged**